

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>sctcrpy.datasets.wu2020 &mdash; sctcrpy  documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script src="https://unpkg.com/@jupyter-widgets/html-manager@^0.18.0/dist/embed-amd.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
  <link rel="stylesheet" href="_static/typehints.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="sctcrpy.datasets.wu2020_3k" href="sctcrpy.datasets.wu2020_3k.html" />
    <link rel="prev" title="sctcrpy.read_tracer" href="sctcrpy.read_tracer.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> sctcrpy
          

          
          </a>

          
            
            
              <div class="version">
                0.0+1add9f4
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <p class="caption"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="introduction.html">Single cell analysis of T cell receptor sequencing in python</a></li>
<li class="toctree-l1"><a class="reference internal" href="features.html">Features</a></li>
<li class="toctree-l1"><a class="reference internal" href="tutorials.html">Tutorials</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="api.html">API</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="api.html#input-output">Input/Output</a></li>
<li class="toctree-l2 current"><a class="reference internal" href="api.html#module-sctcrpy.datasets">Datasets</a><ul class="current">
<li class="toctree-l3 current"><a class="current reference internal" href="#">sctcrpy.datasets.wu2020</a></li>
<li class="toctree-l3"><a class="reference internal" href="sctcrpy.datasets.wu2020_3k.html">sctcrpy.datasets.wu2020_3k</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-sctcrpy.pp">Preprocessing: <code class="docutils literal notranslate"><span class="pre">pp</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#module-sctcrpy.tl">Tools: <code class="docutils literal notranslate"><span class="pre">tl</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="api.html#plotting-pl">Plotting: <code class="docutils literal notranslate"><span class="pre">pl</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="bibliography.html">Bibliography</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">sctcrpy</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
          <li><a href="api.html">API</a> &raquo;</li>
        
      <li>sctcrpy.datasets.wu2020</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            
              <a href="https://github.com/grst/sctcrpy/blob/master/docs/sctcrpy.datasets.wu2020.rst" class="fa fa-github"> Edit on GitHub</a>
            
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  
<style>
/* CSS overrides for sphinx_rtd_theme */

/* 24px margin */
.nbinput.nblast.container,
.nboutput.nblast.container {
    margin-bottom: 19px;  /* padding has already 5px */
}

/* ... except between code cells! */
.nblast.container + .nbinput.container {
    margin-top: -19px;
}

.admonition > p:before {
    margin-right: 4px;  /* make room for the exclamation icon */
}

/* Fix math alignment, see https://github.com/rtfd/sphinx_rtd_theme/pull/686 */
.math {
    text-align: unset;
}
</style>
<div class="section" id="sctcrpy-datasets-wu2020">
<h1>sctcrpy.datasets.wu2020<a class="headerlink" href="#sctcrpy-datasets-wu2020" title="Permalink to this headline">¶</a></h1>
<dl class="function">
<dt id="sctcrpy.datasets.wu2020">
<code class="sig-prename descclassname">sctcrpy.datasets.</code><code class="sig-name descname">wu2020</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#sctcrpy.datasets.wu2020" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the dataset from <a class="reference internal" href="bibliography.html#wu2020" id="id1"><span>[Wu2020]</span></a> as AnnData object.</p>
<p>200k cells, of which 100k have TCRs.</p>
<p>This is how the dataset was processed:</p>
<p><a href="#id2"><span class="problematic" id="id3">``</span></a><a href="#id4"><span class="problematic" id="id5">`</span></a>python
# %load_ext autoreload
# %autoreload 2
import scanpy as sc
import sys</p>
<p>sys.path.append(“..”)
import sctcrpy as st
from multiprocessing import Pool
import os
import pandas as pd
from glob import glob</p>
<p># + language=”bash”
# wget “<a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE139555&amp;format=file">https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE139555&amp;format=file</a>”
# wget “<a class="reference external" href="https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE139555&amp;format=file&amp;file=GSE139555%5Fall%5Fmetadata%2Etxt%2Egz">https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE139555&amp;format=file&amp;file=GSE139555%5Fall%5Fmetadata%2Etxt%2Egz</a>”
# tar xvf GSE139555_raw.tar</p>
<p># + language=”bash”
# cd GSM4143662
# for f in <a href="#id6"><span class="problematic" id="id7">*</span></a>.matrix.mtx.gz; do
#   dirname=${f/.matrix.mtx.gz/}
#   mkdir $dirname
#   mv $dirname.genes.tsv.gz $dirname/genes.tsv.gz
#   mv $dirname.matrix.mtx.gz $dirname/matrix.mtx.gz
#   mv $dirname.barcodes.tsv.gz $dirname/barcodes.tsv.gz
#   mv $dirname.filtered_contig_annotations.csv.gz $dirname/filtered_contig_annotations.csv.gz
#   # fix missing feature type column
#   zcat $dirname/genes.tsv.gz |  awk ‘{print $0 “tGene Expression”}’ | gzip &gt; $dirname/features.tsv.gz
# done
# -</p>
<p>mtx_paths = glob(“GSM4143662/GSM*”)</p>
<p>mtx_paths</p>
<dl class="simple">
<dt>metadata = pd.read_csv(</dt><dd><p>“GSM4143662/GSE139555_all_metadata.txt.gz”, sep=”t”, index_col=0</p>
</dd>
</dl>
<p>)</p>
<p>metadata = metadata[[“ident”, “patient”, “sample”, “source”, “clonotype”]]</p>
<p>metadata = metadata.rename(columns={“clonotype”: “clonotype_orig”})</p>
<p>metadata</p>
<dl>
<dt>def _load_adata(path):</dt><dd><p>sample_id = path.split(“-“)[-1].upper()
obs = metadata.loc[metadata[“sample”] == sample_id, :]
adata = sc.read_10x_mtx(path)
adata_tcr = st.read_10x_vdj_csv(</p>
<blockquote>
<div><p>os.path.join(path, “filtered_contig_annotations.csv.gz”)</p>
</div></blockquote>
<p>)
adata.obs_names = [</p>
<blockquote>
<div><p>“{}_{}”.format(sample_id, barcode) for barcode in adata.obs_names</p>
</div></blockquote>
<p>]
adata_tcr.obs_names = [</p>
<blockquote>
<div><p>“{}_{}”.format(sample_id, barcode) for barcode in adata_tcr.obs_names</p>
</div></blockquote>
<p>]
adata.obs = adata.obs.join(obs, how=”left”)
st.pp.merge_with_tcr(adata, adata_tcr)
return adata</p>
</dd>
</dl>
<p>p = Pool()</p>
<p>adatas = p.map(_load_adata, mtx_paths)</p>
<p>adata = adatas[0].concatenate(adatas[1:])</p>
<p>import numpy as np</p>
<p>np.sum(adata.obs[“has_tcr”])</p>
<p>adata.write_h5ad(“wu2020.h5ad”, compression=”lzf”)</p>
<p>adata.obs</p>
<p><a href="#id8"><span class="problematic" id="id9">``</span></a><a href="#id10"><span class="problematic" id="id11">`</span></a></p>
<dl class="field-list simple">
<dt class="field-odd">Return type</dt>
<dd class="field-odd"><p><span class="annotation terse"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.2.dev5+ge4581ab.d20200127)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a></span><span class="annotation full"><a class="reference external" href="https://anndata.readthedocs.io/en/stable/anndata.AnnData.html#anndata.AnnData" title="(in anndata v0.7.2.dev5+ge4581ab.d20200127)"><code class="xref py py-class docutils literal notranslate"><span class="pre">AnnData</span></code></a></span></p>
</dd>
</dl>
</dd></dl>

</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="sctcrpy.datasets.wu2020_3k.html" class="btn btn-neutral float-right" title="sctcrpy.datasets.wu2020_3k" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="sctcrpy.read_tracer.html" class="btn btn-neutral float-left" title="sctcrpy.read_tracer" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Gregor Sturm, Tamas Szabo.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>