Search.setIndex({docnames:["api","features","index","introduction","references","sctcrpy.pp.merge_with_tcr","sctcrpy.read_10x_vdj","sctcrpy.read_h5ad","sctcrpy.read_tracer","sctcrpy.tl.alpha_diversity","sctcrpy.tl.tcr_dist"],envversion:{"scanpydoc.autosummary_generate_imported":1,"scanpydoc.definition_list_typed_field":1,"scanpydoc.elegant_typehints":1,"scanpydoc.rtd_github_links":1,"sphinx.domains.c":1,"sphinx.domains.changeset":1,"sphinx.domains.citation":1,"sphinx.domains.cpp":1,"sphinx.domains.javascript":1,"sphinx.domains.math":2,"sphinx.domains.python":1,"sphinx.domains.rst":1,"sphinx.domains.std":1,"sphinx.ext.intersphinx":1,scanpydoc:1,sphinx:56},filenames:["api.rst","features.rst","index.rst","introduction.rst","references.rst","sctcrpy.pp.merge_with_tcr.rst","sctcrpy.read_10x_vdj.rst","sctcrpy.read_h5ad.rst","sctcrpy.read_tracer.rst","sctcrpy.tl.alpha_diversity.rst","sctcrpy.tl.tcr_dist.rst"],objects:{"":{sctcrpy:[0,0,0,"-"]},"sctcrpy.pp":{merge_with_tcr:[5,1,1,""]},"sctcrpy.tl":{alpha_diversity:[9,1,1,""],tcr_dist:[10,1,1,""]},sctcrpy:{pl:[0,0,0,"-"],pp:[0,0,0,"-"],read_10x_vdj:[6,1,1,""],read_h5ad:[7,1,1,""],read_tracer:[8,1,1,""],tl:[0,0,0,"-"]}},objnames:{"0":["py","module","Python module"],"1":["py","function","Python function"]},objtypes:{"0":"py:module","1":"py:function"},terms:{"10x":6,"class":7,"default":[5,6,7,10],"int":[7,10],"return":[5,6,7,8,10],"true":[5,6,9],The:5,Used:7,Will:[5,8],account:10,adata:[5,9,10],adata_tcr:5,adata_tx:5,addit:5,align:[4,10],all:[5,8],all_contig_annot:6,alpha:[9,10],analysi:[0,1,2],anndata:[5,6,7,8,9,10],annot:6,api:2,arrai:7,as_spars:7,as_sparse_fmt:7,attribut:7,back:7,batch:5,beta:10,bindings_v2:10,bioinformat:4,bmc:4,bool:[6,7],both:5,cdr3:10,cell:[0,1,2,6,8],cell_id:8,choos:7,chunk:7,chunk_siz:7,column:[5,9],compat:1,comput:10,consumpt:7,contain:8,content:2,contig:6,csr:7,csr_matrix:7,current:10,daili:4,daily2016:[4,10],data:[5,6,7,8],dataset:7,dens:7,directori:8,divers:9,document:3,domin:10,each:[6,8],element:7,everyth:5,exist:5,featur:2,file:[7,8],filenam:7,filter:6,filtered_tcr_seq:8,flavor:9,folder:8,format:7,framework:1,from:[5,6,7,8],fulli:7,gap_extend:10,gap_open:10,gener:8,genom:6,global:4,group:9,h5ad:7,hdf5:7,here:7,high:10,high_confid:6,higher:7,how:5,index:5,inform:8,inplac:[5,9],input:2,instead:7,integr:5,is_cel:6,iter:7,its:7,jeff:4,join:5,json:6,keep:[5,6],kei:9,kwarg:5,left:5,left_index:5,librari:[0,4,10],literal_:7,load:7,local:4,lost:5,matrix:10,mean:7,memori:7,merg:5,mode:7,modifi:[5,7],name:7,need:7,none:[5,7,10],object:[5,6,7,8,9,10],obs:[5,6,8,9],one_to_on:5,onli:[6,7,10],other:5,output:[2,8],pairwis:4,panda:5,paramet:[6,7,8,10],parasail:[4,10],pass:[5,7],path:[6,7,8],perform:[9,10],pkl:8,plot:2,point:7,preprocess:2,python:[0,1,2],read:[6,7,8],receptor:[1,2],refer:2,requir:8,result:8,right_index:5,row:7,sampl:6,save:7,scanpi:1,scipi:7,semi:4,sequenc:[2,4,7,10],shannon:9,simd:4,singl:[0,1,2],size:7,spars:7,sparse_matrix:7,speed:7,spmatrix:7,store:7,str:[6,7,8],subst_mat:10,summar:8,take:10,tcr:[0,1,5,6,8],tcrdist:10,than:5,thi:7,through:[7,10],tool:2,tracer:8,transcriptom:5,type:[5,6,7,8,10],unfortun:8,union:7,until:7,use:9,using:5,valid:5,want:7,when:7,which:9,whole:7,within:9,you:7},titles:["API","Features","Welcome to sctcrpy\u2019s documentation!","Single cell analysis of T cell receptor sequencing in python","References","sctcrpy.pp.merge_with_tcr","sctcrpy.read_10x_vdj","sctcrpy.read_h5ad","sctcrpy.read_tracer","sctcrpy.tl.alpha_diversity","sctcrpy.tl.tcr_dist"],titleterms:{alpha_divers:9,analysi:3,api:0,cell:3,document:2,featur:1,input:0,merge_with_tcr:5,output:0,plot:0,preprocess:0,python:3,read_10x_vdj:6,read_h5ad:7,read_trac:8,receptor:3,refer:4,sctcrpy:[2,5,6,7,8,9,10],sequenc:3,singl:3,tcr_dist:10,tool:0,welcom:2}})